### ISSUE TITLE
Package re-useable parts for use in other projects

### ISSUE THREAD

**Issue by jrbourbeau** [numba/numba-integration-testing#17]
This is a really nice repository, I appreciate the straightforward and modular approach taken for setting up integration tests.

I gave a tour of this repository to some colleagues and there was significant interest in reusing this approach to set up integration testing for other projects (e.g. Dask, NumPy). I wanted to open this issue to see if there is any interest in generalizing `numba-integration-testing` for use in other project. 

For example, one potential approach could be to package reusable functionality in `switchboard.py` (e.g. `execute`, `wget_conda`, `conda_install`, ...) into a separate package (e.g. `inttest` or some other name) that could then be used by other projects. So if Dask wanted to set up integration tests, we would just need to write the appropriate test target subclasses:

```python
# dask-integration-tests.py

from inttest import IntegrationTestTarget, git_ls_remote_tags, execute

class SparseTests(IntegrationTestTarget):

    @property
    def name(self):
        return "sparse"

    @property
    def clone_url(self):
        return "https://github.com/pydata/sparse.git"

    @property
    def target_tag(self):
        return git_ls_remote_tags(self.clone_url)[-1]

    @property
    def conda_dependencies(self):
        return ["pip numpy scipy"]

    def install(self):
        execute("pip install -e .[all]")

    def run_tests(self):
        execute("pytest")
```

and we would be more or less up and running. 

cc @rgommers

**Comment by esc** [comment]
@jrbourbeau thanks for the review and the feedback! I did set out to make it simple, straightforward and self contained in a single python file, deliberately avoiding any use of shell scrips. So, it is nice to know that this is being appreciated!

As for generalization, there is definitive interest from my side to make this re-usable and perhaps expand functionality a little. There are also some sleeping dragons lurking in there that will need to get addressed, here is the stuff that comes to mind immediately:

* 'Activating' the conda-environment is brittle; a simplifying assumption is made that in order to activate an environment, all you need to do is prepend the `PATH` with the `bin` directory of the environment. While this has worked reasonably well, it will not hold as `conda` does modify a few more environment variables than just `PATH`. For example when the `switchboard.py` calls `numba -s` to get Numba to inspect the environment and print what it sees, it will not print the packages from the currently activated environment---but from the base env---because it relies on `CONDA_PREFIX` to detect how to query conda-environments with `conda`. The best solution I have sketched from Python so far is to use `subprocess` with `conda activate`: query the environment from the subprocess e.g. using `env` and then return the modified environment to the original Python process via `stdout`. Then, you compute the difference in environments and thereby work out exactly which environment variables would need to be set to properly activate an environment. As far as I can tell, that should cover it. This is really the only show-stopper and it needs to work reliably.

* Some of idempotency is slightly off, there was some issue with not being inside the correct directory if the repository has been cloned already, I can't remember exactly, but that will eventually arise. This only really matters when running locally to debug/set-up a new target.

* The utilities will need docstrings, and some of the functions might have to be pruned.

* *cough* unit tests *cough* https://github.com/numba/numba-integration-testing/issues/8

* Maybe define a markup language way to declaratively configure targets? YAML would be great but isn't included in the stdlib.

* How would this be shipped if it were re-usable?  know that sharing CI tools, which are often shell scrips, can be tricky so they are often vendored and then get out of sync with the original. In this case however, the advantage is that the `switchboard.py` should run on most versions of Python and doesn't have any dependencies beyond the standard library. So, lets's assume most CI systems will ship with some, maybe older, maybe strangely compiled version of Python and hopefully `pip`. Due to the simplicity of the `switchboard.py` such a Python would totally suffice and the first step in any integration-test would be to bootstrap with `pip inttest` and then hand off control to `switchboard.py` (or whatever the entry point will  be called) bootstrap `minconda` and do the rest of the dance from there.

I think it would be good to establish a minimal feature set and a target secondary project to run integration-tests for as that would give us a roadmap on how to proceed. Most of the work will be going from a single project to two projects. Once that is done, scaling  to `n` shouldn't be too hard.

**Comment by jrbourbeau** [comment]
Thanks for the thoughtful reply @esc. Glad to hear there's interest on your side. I agree with all your points to address.

- To your point about activating the testing conda environment, we may be able to make use of the `conda run` command. It will run an executable from inside a specified conda environment. For example, I have a `numba-dev` conda environment on my laptop for working on Numba, and executing `conda run --name numba-dev numba -s` from the base conda environment displays info about the `numba-dev` environment. Note that `conda run` is marked as experimental (see `conda run --help`), but it seems to work and this looks like a good use case for it. I'm curious what you think about this. 

- How would this be shipped if it were re-usable? Having example CI configuration files for Travis, CircleCI, and Azure would be a good start. As you mentioned, `switchboard.py` should be able to do most of the heavy lifting. If we decide to go a little further, one option could be to make a [cookiecutter template](https://github.com/cookiecutter/cookiecutter) to set up some basics (e.g. project metadata, CI configurations, a basic `switchboard.py` with instructions on how to add new targets). But I think having example CI configuration files is a good start. Definitely open to suggestions on how this could be shipped 


For a minimal feature set, I think the features already implemented in `switchboard.py` would be valuable to others and could serve a minimal set. I'm biased, but would be in favor of targeting Dask as a secondary project : )

**Comment by esc** [comment]
@jrbourbeau : I think `conda run` is worth trying out, it may do exactly what we need. I thought about this some more and the distribution question can be deferred. If we only have two targets, Numba and Dask, we only need to solve the problem for these two (and I think I will go down the install-from-pypi route first). My plan of action is to setup a new repo, put the reusable bits in there somehow and then try to refactor `numba-integration-testing` in such a way as to use the code from this new repo. I'll do this refactoring in a branch of `numba-integration-testing` and ping you as soon as it is in a reasonable state so you can guinea pig this for me on Dask. Once that works for Numba and Dask I can merge the branch and we can look for other projects that might want to adopt this. Sound like a plan? Now I only need a name...

**Comment by jrbourbeau** [comment]
> My plan of action is to setup a new repo, put the reusable bits in there somehow

Sounds like a good way forward. I'm happy to help out with this, if you want to link to the new repo

> Now I only need a name...

Naming is hard. A few options that come to mind are `pyinttest` (which is a little awkward) or `inttest`...but those aren't very creative. Do you have any thoughts on a name?

Btw, I'm at SciPy this week if you'd like to chat about this (or anything else)

**Comment by esc** [comment]
@jrbourbeau I believe this issue can now be closed. TexasBBQ has been running the Numba integration tests for over a year now. It comes with it's own test suite and everything so I think this went well. Thanks again for suggesting this and your efforts in helping to move this forward. BTW: how are the Dask tests going? 😘
